src/hyperscan.rs

```diff
@@ -0,0 +1,299 @@
+            expression: pattern_list.get(&P_IP).unwrap().clone(),
+            id: P_IP.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::EMAIL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_EMAIL).unwrap().clone(),
+            id: P_EMAIL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::DATE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_DATE).unwrap().clone(),
+            id: P_DATE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::QUOTED) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_QUOTED).unwrap().clone(),
+            id: P_QUOTED.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::URL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_URL).unwrap().clone(),
+            id: P_URL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::PHONE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_PHONE).unwrap().clone(),
+            id: P_PHONE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::USER_AGENT) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_USER_AGENT).unwrap().clone(),
+            id: P_USER_AGENT.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+
+    let res_db: Result<BlockDatabase, _> = patterns.build();
+
+    match res_db {
+        Ok(db) => Some(db),
+        Err(_) => None,
+    }
+}
+
+pub struct HSPatternMatch {
+    pub pattern_id: u32,
+    pub from: u64,
+    pub to: u64,
+}
+
+pub struct HSLineScanner<'a> {
+    //    pub inner: Arc<Mutex<LineScannerData>>,
+    pub lines: &'a Vec<String>,
+    pub line_matches: HashMap<u16, HashMap<u16, Vec<HSPatternMatch>>>,
+}
+
+pub type HSPatternMatchResults =
+    Arc<RwLock<HashMap<u16, RwLock<HashMap<u16, Vec<HSPatternMatch>>>>>>;
+
+impl<'a> HSLineScanner<'a> {
+    pub fn new(lines: &Vec<String>) -> HSLineScanner {
+        HSLineScanner {
+            lines: lines,
+            line_matches: HashMap::new(),
+        }
+    }
+
+    pub fn scan(&mut self, db: &mut BlockDatabase) -> HSPatternMatchResults {
+        let now = Instant::now();
+
+        let line_total = self.lines.len();
+        let scratch = db.alloc().unwrap();
+
+        let pattern_match_results: HSPatternMatchResults = Arc::new(RwLock::new(HashMap::new()));
```

Comment:
Why do we need a lock here -- it is not multi-threaded, right?

Reply:
Because we send this object to the callback, Rust borrow checker won't allow me to pass ownership of the structure multiple times, so I need to protected behind a lock and count how many times it's been referenced

Reply:
Yeah, I see -- C-style callbacks like this are a problem for Rust (as it cannot infer that the callback will not be called multi-threaded ...).

Can we do some "unsafe" stuff to get rid of it? (As we know that Hyperscan is single threaded and will not fire overlapping callbacks).

Reply:
Each scan invocation should be able to run on it's own thread, what we can do to make the callbacks faster is to send them into a `Sink` or channel so the next callback can be called regardless of how the last callback result is getting processed